Modify rule S2637(cfamily): CPP-5602 allow assigning nullptr to params (#4235)
SonarSource/rspec: rules/S2637/cfamily/rule.adoc

#include <string.h>

__attribute__((returns_nonnull)) int *

> make_array_copy(int *src, size_t len)__attribute__((nonnull(1)));

Reviewer: I'm not familiar with this syntax, but do we actually start at `1` or is it `0` to denote the very first function parameter?

Author: We start from 1, see [compiler explorer](https://compiler-explorer.com/z/EaqajEf77)

Reviewer: Thanks for clarifying. It still looks weird, though 😅

> make_array_copy(int *src, size_t len)__attribute__((nonnull(1)));

Reviewer: In one of the other usages of this attribute, we do have an additional space ` ` between the function's parameter list and the `__attribute__`.
```suggestion
make_array_copy(int *src, size_t len) __attribute__((nonnull(1)));
```

Failing to adhere to the attribute may introduce serious program errors.
In particular, the compiler does not enforce that values marked as `nonnull` are indeed non-null at runtime; it is the developers' responsibility to adhere to the attribute.
These values are typically _not_ null-checked before use.

> Passing null (i.e., `NULL`, `0` or `nullptr`) argument to a parameter as `nonnull`
> or returning null from a function marked as `returns_nonnull` is hence likely to cause a null-pointer dereference.

Reviewer: ```suggestion
Passing null (i.e., `NULL`, `0` or `nullptr`) as an argument to a parameter that is marked as `nonnull`
or returning null from a function marked as `returns_nonnull` is, hence, likely to cause a null-pointer dereference.
```

Compilers may even apply optimizations based on this attribute and might, for instance, _remove_ an explicit null-check if the parameter is declared as `nonnull` -- even in code outside of the function with the attribute.

Note that the `nonnull` attribute is a GNU extension (see https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-nonnull-function-attribute[nonnull] and https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-returns_005fnonnull-function-attribute[returns_nonnull]) which many compiler vendors have implemented.

Note that it is allowed to _assign_ null to a parameter marked as `nonnull`.

> This attribute specifies only the function call contract
> and does not control the evolution of the parameter variable.

Reviewer: ```suggestion
This attribute is only concerned with the function call contract
and does not control the evolution of the parameter variable.
```

For example, a linked-list search could be implemented as follows:

[source,cpp]
----
struct List {
  int value;

>   List* next; // NULL for a tail node.

Reviewer: ```suggestion
  List* next; // nullptr for a tail node.
```
But I'm also fine with `NULL`.

Besides affecting the application's availability, null-pointer dereferences may lead to malicious code execution, in rare circumstances.
If null is equivalent to the 0x0 memory address that can be accessed by privileged code, writing, and reading memory is possible, which compromises the integrity and confidentiality of the application.

Because compilers may apply optimizations based on the `nonnull` attribute, not respecting `nonnull` can also introduce more complex bugs such as resource leaks or infinite recursion as indicated in the following code snippet:

[source,cpp]
----
struct List {
  int value;

>   List *next; // NULL for a tail node.

Reviewer: ?
```suggestion
  List *next; // nullptr for a tail node.
```
But I'm fine with `NULL` here, too.

>   List *next; // NULL for a tail node.

Reviewer: Both, `List *next` and `List* next` are used. Can we settle on one style? Preferably `List *next`.
Let's not add `List * next` 😁

Author: fixed in d1844af83e

};

> size_t len(List *n) __attribute__((nonnull));

Reviewer: We don't need an index here? (I really don't know, since I always use the attribute directly next to the argument.)

Author: having no index means it applies to all parameters.
I think it is more readable for 1-argument function, but should not be used if the function takes multiple arguments.

Reviewer: Yeah, it's more readable in that case. Cool.

> size_t len(List *n) {
>   if (!n) return 0; // Impossible branch according to the attribute
>   return 1 + len(n->next);

Reviewer: Let's be consistent with naming our pointer-to-list-typed parameters.
```suggestion
size_t len(List *l) {
  if (!l) return 0; // Impossible branch according to the attribute
  return 1 + len(l->next);
```

}
----

== How to fix it

> Ensure not to pass null values when `nonnull` arguments are expected and not to return a null value when a function marked as `returns_nonnull`.

Reviewer: ```suggestion
Ensure not to pass null values when `nonnull` arguments are expected and not to return a null value when a function is marked as `returns_nonnull`.
```

[source,cpp,diff-id=1,diff-type=noncompliant]
----

> int* foo(int* x) __attribute__((nonnull));

Reviewer: Does this syntax mean that any pointer input parameter and return value might not be null?

Author: Only the former. It is orthogonal with `returns_nonnull`.
I removed `returns_nonnull` because we have another example for it, and it does not change anything in this example

Reviewer: Makes sense. Nice!

[source,cpp,diff-id=1,diff-type=compliant]
----

> int* foo(int* x) __attribute__((nonnull));

Reviewer: Same question.

Reviewer: I'm reading this on Github, but I currently fail to see the difference between the non-compliant and compliant code snippets. Could you point out the difference to me?

Reviewer: Thanks for clarifying.

> int* foo(int* x) {

Reviewer: And can we have the `*` always next to the variable/and function (and not next to the type)? Or, in case you like the other spelling better, we should at least be consistent.

Author: fixed in d1844af83e

Reviewer: Awesome 🎉
